ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala

```diff
@@ -0,0 +1,104 @@
+package pl.touk.nussknacker.ui.notifications
+
+import akka.actor.ActorRef
+import akka.pattern.ask
+import akka.util.Timeout
+import com.typesafe.scalalogging.LazyLogging
+import pl.touk.nussknacker.engine.api.process.{ProcessId, ProcessName}
+import pl.touk.nussknacker.ui.listener.ProcessChangeEvent.{OnDeployActionFailed, OnDeployActionSuccess}
+import pl.touk.nussknacker.ui.listener.{ProcessChangeEvent, ProcessChangeListener, User}
+import pl.touk.nussknacker.ui.process.deployment.{DeployInfo, DeploymentActionType, DeploymentStatus, DeploymentStatusResponse}
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+
+import java.time.temporal.ChronoUnit
+import java.time.{Clock, Instant}
+import java.util.UUID
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+
+case class NotificationConfig(duration: FiniteDuration)
+
+class NotificationsListener(config: NotificationConfig,
+                            fetchName: ProcessId => Future[Option[ProcessName]],
+                            clock: Clock = Clock.systemUTC()) extends ProcessChangeListener with LazyLogging {
+
+  //not too efficient, but we don't expect too much data...
+  private val data: ArrayBuffer[NotificationEvent] = ArrayBuffer()
+
+  override def handle(event: ProcessChangeEvent)(implicit ec: ExecutionContext, user: User): Unit = {
+    val now = Instant.now(clock)
+    fetchName(event.processId).onComplete {
+      case Failure(NonFatal(e)) => logger.error(s"Failed to retrieve scenario name for id: ${event.processId}", e)
+      case Success(None) => logger.error(s"Failed to retrieve scenario name for id: ${event.processId}")
+      case Success(Some(scenarioName)) => synchronized {
+        data.append(NotificationEvent(UUID.randomUUID().toString, event, now, user, scenarioName))
+      }
+    }
+    filterOldNotifications(now)
+  }
+
+  private def filterOldNotifications(now: Instant): Unit = synchronized {
+    data.zipWithIndex.filter(_._1.date.isBefore(now.minus(config.duration.toMillis, ChronoUnit.MILLIS))).foreach(i => data.remove(i._2))
```

Comment:
will it work on cloud with different sever and client timezones ?

Reply:
we use Instant here, no time zones.